server/lib/s3backup/__init__.py

```diff
@@ -0,0 +1,106 @@
+from configparser import ConfigParser
+
+
+class Entry(object):
+    """
+    An Entry object consists of a name (of an S3 object) and the MD5 value of that object. It is used to create an object with the name and MD5 value, so that it can be compared with other similar objects.
+    """
+    def __init__(self, name, md5):
+        self.name = name
+        self.md5 = md5
+
+    def __eq__(self, other):
+        return self.name == other.name and self.md5 == other.md5
+
+
+class S3Connector(object):
+    def __init__(self, access_key_id, secret_access_key, endpoint_url, bucket_name):
+        self.s3connector = boto3.client('s3',
+                                        aws_access_key_id='{}'.format(
+                                            access_key_id),
+                                        aws_secret_access_key='{}'.format(
+                                            secret_access_key),
+                                        endpoint_url='{}'.format(endpoint_url))
+
+    def ts(self, *args):
+        return time.strftime("%Y-%m-%dT%H:%M:%S-%Z", time.localtime())
+
+    def list_objects(self, Bucket):
+        return self.s3connector.list_objects_v2(Bucket=Bucket)
+
+    def put_object(self, Bucket, Key, Body, ContentMD5):
+        return self.s3connector.put_object(Bucket=Bucket, Key=Key,
+                Body=Body, ContentMD5=ContentMD5)
+
+
+class MockS3Connector(object):
+    """
+    The mock object is used for unit testing. It provides a "connector" to the backend service that is implemented using the local filesystem, rather than dispatching to the real S3 backend service.
+    """
+
+    def __init__(self, access_key_id, secret_access_key, endpoint_url, bucket_name):
+        self.path = endpoint_url
+        self.bucket_name = bucket_name
+
+    def ts(self, *args):
+        return "1900-01-01T00:00:00-UTC"
+
+    def list_objects(self, Bucket):
+        ob_dict = {}
+        result_list = glob.glob(os.path.join(
+            self.path, Bucket, "*.tar.xz"))
+        result_name_list = []
+        for i in result_list:
+            with open(i, 'rb') as f:
+                data = f.read()
+                md5 = hashlib.md5(data).hexdigest()
+            result_name_list.append({'ETag': '"{}"'.format(md5), 'Key': i})
+        ob_dict['Contents'] = result_name_list
+        ob_dict['ResponseMetadata'] = {'HTTPStatusCode': 400}
+        return ob_dict
+
+    def put_object(self, Bucket=None, Key=None, Body=None, ContentMD5=None):
+        md5_hex_value = hashlib.md5(Body).hexdigest()
+        md5_base64_value = (base64.b64encode(
+                        bytes.fromhex(md5_hex_value))).decode()
+        if md5_base64_value == ContentMD5:
+            with open('{}/{}/{}.tar.xz'.format(self.path, self.bucket_name, Key), 'wb') as f:
+                f.write(Body)
+
+
+class S3Config(object):
+    def __init__(self, config):
+        try:
+            debug_unittest = config.get('pbench-server-backup', 'debug_unittest')
```

Comment:
You should get this from the `[pbench-server]` section as noted above for `pbench-server.cfg.example`.